src/os_posix/os_dir.c

```diff
@@ -64,17 +90,47 @@ __directory_list_worker(WT_FILE_SYSTEM *file_system, WT_SESSION *wt_session, con
         if (single)
             break;
     }
-
+    /*
+     * Reading the directory returns NULL on failure or reaching the end of the list. Record a
+     * special message if readdir failed.
+     */
+    if (errno != 0) {
+        __wt_epoch(session, &ts);
+        WT_ERR(__wt_buf_fmt(session, readerrmsg,
+          "[%" PRIuMAX ":%" PRIuMAX "] readdir failed errno %d (%s) dir fd %d",
+          (uintmax_t)ts.tv_sec, (uintmax_t)ts.tv_nsec / WT_THOUSAND, errno,
+          __wt_strerror(session, errno, NULL, 0), dirfd(dirp)));
+        ret = errno;
```

Comment:
Are we sure that none of the functions above change errno? It would be safer to save its value first, at the beginning of the branch.

Reply:
Fixed.